internal/provider/role_resource.go

```diff
@@ -0,0 +1,238 @@
+	resp.Diagnostics.Append(d...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, result)...)
+}
+
+func (r *roleResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	// Get current state
+	var state roleResourceModel
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Get refreshed resource state from API
+	role, err := r.ClientV2.Role.GetRole(ctx, state.Id.ValueString(), nil)
+	if err != nil {
+		if errors.Is(err, client.ErrNotFound) {
+			resp.State.RemoveResource(ctx)
+			return
+		}
+		resp.Diagnostics.AddError("Error retrieving role", err.Error())
+		return
+	}
+
+	result, d := roleResourceModelFromAPI(ctx, role)
+	resp.Diagnostics.Append(d...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, result)...)
+}
+
+func (r *roleResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	var plan, state roleResourceModel
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	opts := schemas.RoleRequest{}
+
+	if !plan.Name.Equal(state.Name) {
+		opts.Attributes.Name = value.Set(plan.Name.ValueString())
+	}
+
+	if !plan.Description.Equal(state.Description) {
+		opts.Attributes.Description = framework.SetIfKnownString(plan.Description)
+	}
+
+	if !plan.Permissions.Equal(state.Permissions) && !plan.Permissions.IsUnknown() && !plan.Permissions.IsNull() {
+		var permissionIDs []string
+		resp.Diagnostics.Append(plan.Permissions.ElementsAs(ctx, &permissionIDs, false)...)
+
+		permissions := make([]schemas.Permission, len(permissionIDs))
+		for i, id := range permissionIDs {
+			permissions[i] = schemas.Permission{ID: id}
+		}
+
+		opts.Relationships.Permissions = value.Set(permissions)
+	}
+
+	// Update existing resource
+	role, err := r.ClientV2.Role.UpdateRole(ctx, plan.Id.ValueString(), &opts, nil)
+	if err != nil {
+		resp.Diagnostics.AddError("Error updating role", err.Error())
+		return
+	}
+
+	result, d := roleResourceModelFromAPI(ctx, role)
+	resp.Diagnostics.Append(d...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, result)...)
+}
+
+func (r *roleResource) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
+	var state roleResourceModel
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	err := r.ClientV2.Role.DeleteRole(ctx, state.Id.ValueString())
+	if err != nil && !errors.Is(err, client.ErrNotFound) {
+		resp.Diagnostics.AddError("Error deleting role", err.Error())
+		return
+	}
+}
+
+func (r *roleResource) ImportState(ctx context.Context, req resource.ImportStateRequest, resp *resource.ImportStateResponse) {
```

Comment:
@soltysss Here is practically the only fix made; the rest of the PR is the adoption of a new API client and refactoring of the role resource to the Framework.